[RISCV] Select mask operands as virtual registers and eliminate uses of vmv0

lukel97 opened the pull request:
This is another attempt at #88496 to keep mask operands in SSA after instruction selection.

Previously we selected the mask operands into vmv0, a singleton register class with exactly one register, V0.

But the register allocator doesn't really support singleton register classes and we ran into errors like "ran out of registers during register allocation in function".

This avoids this by introducing a pass just before register allocation that converts any use of vmv0 to a copy to $v0, i.e. what isel currently does today.

That way the register allocator doesn't need to deal with the singleton register class, but we get the benefits of having the mask registers in SSA throughout the backend:

- This allows RISCVVLOptimizer to reduce the VLs of instructions that define mask registers
- It enables CSE and code sinking in more places
- It removes the need to peek through mask copies in RISCVISelDAGToDAG and keep track of V0 defs in RISCVVectorPeephole

This patch initially eliminates uses of vmv0s after RISCVVectorPeephole to keep the diff to a minimum, and a follow up patch will move it past the other MachineInstr SSA passes.

Note that it doesn't try to remove any defs of vmv0 as we shouldn't have any instructions that have any vmv0 outputs.

As a further follow up, we can move the elimination pass to after phi elimination and outside of SSA, which would unblock the pre-RA scheduler around masked pseudos. This might also help the issue that RISCVVectorMaskDAGMutation tries to solve.

lukel97 committed:
[RISCV] Select mask operands as virtual registers and eliminate vmv0

This is another attempt at #88496 to keep mask operands in SSA after instruction selection.

Previously we selected the mask operands into vmv0, a singleton register class with exactly one register, V0.

But the register allocator doesn't really support singleton register classes and we ran into errors like "ran out of registers during register allocation in function".

This avoids this by introducing a pass just before register allocation that converts any use of vmv0 to a copy to $v0, i.e. what isel currently does today.

That way the register allocator doesn't need to deal with the singleton register class, but get the benefits of having the mask registers in SSA throughout the backend:

- This allows RISCVVLOptimizer to reduce the VLs of instructions that define mask registers
- It enables CSE and code sinking in more places
- It removes the need to peek through mask copies in RISCVISelDAGToDAG and keep track of V0 defs in RISCVVectorPeephole

As a follow up, we can move the elimination pass to after phi elimination and outside of SSA, which would unblock the pre-RA scheduler around masked pseudos. This might also help the issue that RISCVVectorMaskDAGMutation tries to solve.

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp.ll:
```diff
@@ -3916,10 +3916,9 @@ define void @trunc_v6bf16(ptr %x) {
 ; CHECK-NEXT:    vsetivli zero, 8, e16, m1, ta, ma
 ; CHECK-NEXT:    vfwcvtbf16.f.f.v v10, v8
-; CHECK-NEXT:    vsetvli zero, zero, e32, m2, ta, ma
+; CHECK-NEXT:    vsetivli zero, 6, e32, m2, ta, ma
```
This is from RISCVVLOptimizer being able to propagate VLs through mask users now

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/commutable.ll:
```diff
@@ -26,10 +26,9 @@ define <vscale x 1 x i64> @commutable_vadd_vv_masked(<vscale x 1 x i64> %0, <vsc
 ; CHECK-NEXT:    vadd.vv v8, v8, v9, v0.t
 ; CHECK-NEXT:    vsetvli a0, zero, e64, m1, ta, ma
-; CHECK-NEXT:    vadd.vv v8, v10, v8
+; CHECK-NEXT:    vadd.vv v8, v8, v8
```
This is an example of CSE

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/vector-extract-last-active.ll:
```diff
@@ -5,13 +5,14 @@
 ; CHECK-NEXT:  # %bb.1:
+; CHECK-NEXT:    vmv.v.i v9, 0
+; CHECK-NEXT:    vsetvli zero, zero, e8, m1, ta, mu
+; CHECK-NEXT:    vid.v v9, v0.t
```
This is an example of code sinking

wangpc-pp reviewed (approved):
Thanks! I just had a quick scan and I like this method to work around the limitation. 
LGTM but please wait for another approval.

topperc commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,154 @@
+  for (MachineBasicBlock *MBB : RPOT) {
+    bool V0Clobbered = V0ClobberedOnEntry.contains(MBB);
+    for (MachineInstr &MI : *MBB) {
+      assert(!(MI.readsRegister(RISCV::V0, TRI) && V0Clobbered));
```
Needs error message

preames commented on llvm/test/CodeGen/RISCV/rvv/vector-extract-last-active.ll:
```diff
@@ -124,12 +125,12 @@ define float @extract_last_float(<4 x float> %data, <4 x i32> %mask, float %pass
 ; CHECK-NEXT:  # %bb.1:
+; CHECK-NEXT:    vsetvli zero, zero, e8, mf4, ta, mu
+; CHECK-NEXT:    vmv.v.i v9, 0
+; CHECK-NEXT:    vid.v v9, v0.t
```
Complete aside: We could use a start value of -1 here, use a vredmaxs instead, and move the conditional branch to be after the reduction, and guard only the slidedown and extract.  Doing so would remove a vcpop.  This assumes that the passthru case is exceedingly rare.  

Actually, we could replace the slidedown entirely with another masked reduction (mask being a comparison of the vid to the result of the max), and feed the passthru through as the start value for an entirely branchless sequence.  Not sure the required vrgather.vi is worth it there though.

topperc commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,154 @@
+  // For any instruction with a vmv0 operand, replace it with a copy to v0.
+  for (MachineBasicBlock &MBB : MF) {
+    for (MachineInstr &MI : MBB) {
+      // An instruction should only have one or zero vmv0 operands.
```
Move this comment into a message on the assert?

topperc commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,154 @@
+        if (MO.isReg() && MO.getReg().isVirtual() &&
+            MRI.getRegClass(MO.getReg()) == &RISCV::VMV0RegClass) {
+          MRI.recomputeRegClass(MO.getReg());
+          assert(MRI.getRegClass(MO.getReg()) != &RISCV::VMV0RegClass ||
```
message

preames commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,154 @@
+    bool V0Clobbered = V0ClobberedOnEntry.contains(MBB);
+    for (MachineInstr &MI : *MBB) {
+      assert(!(MI.readsRegister(RISCV::V0, TRI) && V0Clobbered));
+      if (MI.modifiesRegister(RISCV::V0, TRI))
```
I'm not sure this invariant holds.  Consider the following:
// inline asm that sets V0
// call to some VP intrinsic with a mask
// inline asm which reads V0

This sequence is fully UB, but I think your assert might spurious fail in this case.

lukel97 replied:
I added a test case for this in vmv0-elimination.ll, but couldn't get the assertion to trigger.

It looks like any inline asm that reads v0 (i.e. has a v0 input constraint: the vm constraint isn't affected here) will end up emitting a glued COPY to $v0 before it:

```
bb.0 (%ir-block.0):
  liveins: $v8, $v9, $v0, $x10
  %3:gpr = COPY $x10
  %2:vr = COPY $v0
  %1:vr = COPY $v9
  %0:vr = COPY $v8
  %4:vr = COPY %0:vr
  %5:vr = COPY %1:vr
  INLINEASM &"vadd.vv $0, $1, $2" [attdialect], $0:[regdef], implicit-def $v0, $1:[reguse:VR], %4:vr, $2:[reguse:VR], %5:vr
  %8:vr = COPY $v0
  %10:vmv0 = COPY %2:vr
  %9:vrnov0 = PseudoVADD_VV_M1_MASK $noreg(tied-def 0), %0:vr, %1:vr, %10:vmv0, -1, 6, 0
  PseudoVSE64_V_M1 killed %9:vrnov0, %3:gpr, -1, 6 :: (store (<vscale x 1 x s64>) into %ir.p)
  %7:vr = COPY %0:vr
  $v0 = COPY %8:vr
  INLINEASM &"vadd.vv $0, $1, $2" [attdialect], $0:[regdef:VR], def %6:vr, $1:[reguse:VR], %7:vr, $2:[reguse], $v0
  $v8 = COPY %6:vr
  PseudoRET implicit $v8
```

So any inline asm which reads V0 shouldn't be clobbered. Should we maintain this stricter invariant then?

preames reviewed (commented):
As we talked about last night, very supportive of this direction.  I think this approach works out.

One suggestion: can we stage this into two commits to make the diffs easier to understand?  If you do the new pass immediately after ISEL, I think you get something which is much closer to NFC.  This lets us land the majority of the cleanup  Then you can do a separate change which moves it before regalloc, and we can focus on the interaction with all the passes being moved over.

topperc commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,154 @@
+  };
+
+#ifndef NDEBUG
+  // Assert that we won't clobber any existing reads of V0 where we need to
```
Can we use the livein list on the basic blocks?

lukel97 replied:
I've replaced the map with an assertion that none of the successors have v0 as a live in if v0 is clobbered

arsenm commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,154 @@
+// register in the singleton vmv0 register class instead of copying them to $v0
+// straight away, to make optimizing masks easier.
+//
+// However the register allocator struggles with singleton register classes and
```
This is a hand wavy and poor explanation for the problem and makes it sound like a workaround. The problem is primarily with every other piece of code that can now produce situations where multiple values must be live at the same time

preames replied:
To my knowledge, it is a workaround?  I don't understand your second sentence.  All the words individually make sense, I just don't get the connection to the point.  I've never really understood why 3 virtual registers with overlapping live ranges in a 2 element register class is fine, but 2 overlapping virtual registers in a 1 element register class isn't.  

Also, do you have better wording to use here?  Or is this more than a request to reword a comment?

lukel97 replied:
> I've never really understood why 3 virtual registers with overlapping live ranges in a 2 element register class is fine, but 2 overlapping virtual registers in a 1 element register class isn't.

Maybe this is just a limitation of how GreedyRegAlloc implements spilling, i.e. it requires at least two registers in a class?

One way to think of RISCVVMV0Elimination is that it "spills" vmv0 register classes to vr register classes by inserting copies and inflating the register class. That way we're removing any overlapping live ranges.

arsenm replied:
The problem is if you have more than 1 value that need to be live at the same point, as in instruction.

```
  %0:my_rc = DEF0
  %1:my_rc = DEF1
  SOME_INST implicit %0, implicit %1
```
Is the kind of case that will fail. Other cases where they don't need to be live at the same point should work

lukel97 replied:
Off the top of my head the reason why we were ending up with instructions with two separate uses of vmv0 was because copies were getting coalesced into the vmv0 class. There shouldn't be any instructions with two vmv0 operands.

It might be worth trying to change this pass to inflate any unnecessary vmv0 register classes so we avoid this situation? Ideally in a separate PR so we minimise the test diff from isel's current behaviour

preames replied:
Luke, I'm not clear what you mean by "this pass".  It sounds like you're referring to the VMV0 elimination pass from this patch, but that doesn't seem to make sense in context?  Were you maybe referring to Register Coalescing?

lukel97 replied:
My idea is that RISCVVMV0Elimination sits just before register allocation, after register coalescer. But instead of emitting copies to V0, it finds instructions with two or more distinct uses of VMV0.

Only one of the uses should need to be in VMV0, I.e the mask operand.

For the other uses, it would either inflate the uses to VR, or if not possible emit a COPY to VR. 

That way each instruction would have at most one use of VMV0, and this avoids the scenario that @arsenm was mentioning, which should be allocatable. 

I'm suggesting we could explore it as a follow up to this patch.

lukel97 committed:
Move pass to just after RISCVVectorPeephole

lukel97 committed:
Address review comments
- Peek through copies to match isel better, remove most of test diff
- Use liveins list to check for clobbers across blocks
- Move messages into asserts
- Add test for inline asm use
- Assert no subregster index

lukel97 commented:
> One suggestion: can we stage this into two commits to make the diffs easier to understand? If you do the new pass immediately after ISEL, I think you get something which is much closer to NFC. This lets us land the majority of the cleanup Then you can do a separate change which moves it before regalloc, and we can focus on the interaction with all the passes being moved over.

I've shuffled it back to just after RISCVVectorPeephole (so we can remove the V0 def map in it), and peeked through the first COPY to match the behaviour of SelectionDAG, most of the test diffs should be gone now. 

The one place I couldn't get rid of the diffs are in the nearbyint tests. The changes are due to how the VFROUND pseudos are expanded in such a way that two pseudos might share the same $v0 COPY, instead of having one copy per pseudo, e.g.:

```
$v0 = COPY %17:vr
%41:vrm8nov0 = PseudoVFCVT_X_F_V_M8_MASK %20:vrm8nov0(tied-def 0), %0:vrm8, $v0, 7, %12:gprnox0, 6, 3, implicit $frm
%19:vrm8nov0 = PseudoVFCVT_F_X_V_M8_E64_MASK %20:vrm8nov0(tied-def 0), %41:vrm8nov0, $v0, 7, %12:gprnox0, 6, 3, implicit $frm
```

It's possible to remove redundant copies to $v0 in this pass but it ends up generating more test diffs, so shelving that idea for later.

lukel97 commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -599,6 +602,10 @@ void RISCVPassConfig::addMachineSSAOptimization() {
 
 void RISCVPassConfig::addPreRegAlloc() {
+  // TODO: Move this as late as possible before regalloc
+  if (TM->getOptLevel() == CodeGenOptLevel::None)
```
At -O0 we don't run `addMachineSSAOptimization` so we still need to eliminate them here

preames commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,171 @@
+                 Src.isVirtual() && "vmv0 use in unexpected form");
+
+          // Peek through a single copy to match what isel does.
+          MachineInstr *SrcMI = MRI.getVRegDef(Src);
```
SrcMI isn't used after this, and can be faulted into the if-clause "if (def; condition)"

preames commented on llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp:
```diff
@@ -0,0 +1,171 @@
+                 MI.isInlineAsm() ||
+                 MRI.getVRegDef(MO.getReg())->isInlineAsm() &&
+                     "Non-inline-asm use of vmv0 left behind");
+          MadeChange = true;
```
This update is redundant.  At this point, the value is unconditionally true at the return below.  Please remove, and change the return to "return true;" below.

preames reviewed (approved):
LGTM w/comments addressed.

Though, I'd encourage you to wait on review from @topperc .  He's better at spotting subtle problems in this type of thing than I am.

lukel97 committed:
Remove madeChange assignment, fold def into if statement, clarify comment about how the register coalescer produces unallocatable instructions

wangpc-pp reviewed (approved):
LGTM. Thanks! This is GREAT!

topperc reviewed (approved):
LGTM

topperc commented:
I went back and looked at the `early-clobber %res:vr = PseudoVMSEQ_VV_M2_MASK %mask:vmv0(tied-def 0), ..., %mask:vmv0, ...` mentioned in https://github.com/llvm/llvm-project/pull/88496

I wonder if we can fix that by using VRNoV0 on the passthru operand. Due to the early-clobber we never end up using V0 for that operand today anyway. I have a ticket for that in our internal database from 2+ years ago:

> Masked compares can't use V0 as destination if mask and maskedoff are the same. This means we can’t directly implement this sequence from the V spec
> 
> Example of implementing isgreater():
>
>vmfeq.vv v0, va, va
>vmfeq.vv v1, vb, vb
> [vmand.mm](http://vmand.mm/) v0, v0, v1
> vmfgt.vv v0, va, vb, v0.t  #  so only set flags on ordered values.
> 
> I tried to add TIED instructions to use the same tied operand for both. Unfortunately, if you put V0 twice in the isel pattern, that does not cause the pattern match to check that they are both the same value. This functionality only works if you have a named operand in the pattern like “$rs2”. I believe we’ll need custom isel to handle this. That’s a lot more complicated since we need to pick the opcode by LMUL.
>
> A post-isel peephole might also work.
